File: leetcode/0202-happy-number/solution.py
class Solution:
    def isHappy(self, n: int) -> bool:
        """
        RUNTIME: 34 ms (93.99%)
        MEMORY: 13.8 MB (62.35%)
        """
        if n == 1: return True
        hashset = {}
        while n != 1:
            if n < 10:
                n = n * n
            else:
                sum = 0
                while True:
                    sum += (n % 10) * (n % 10)
                    if n // 10 == 0:
                        break
                    n = n // 10
                n = sum
                if n == 1:
                    return True
            if n not in hashset:
                hashset[n] = 1
            else:
                return False

a = Solution()
print(a.isHappy(19))

File: leetcode/0202-happy-number/solution.cpp
class Solution {
public:
    /**
     * RUNTIME: 0 ms (100%)
     * MEMORY: 6.2 MB (39.70%)
    */
    bool isHappy(int n) {
        if (n == 1) { return true; }
        std::unordered_map<int, int> hashset;
        int sum;
        while (true) {
            if (n < 10) {
                n *= n;
            }
            else {
                sum = 0;
                while (true) {
                    unsigned tmp = (n % 10) * (n % 10);
                    sum += tmp;
                    if (n / 10 == 0) {
                        break;
                    }
                    n = n / 10;
                }
                n = sum;
                if (n == 1) { return true; }
            }
            if (hashset.find(n) == hashset.end() ) {
                hashset[n] = 1;
            }
            else {
                return false;
            }
        }
        return false;
    }
};

int main() {
    Solution a;
    std::cout << a.isHappy(19);
    return -1;
}